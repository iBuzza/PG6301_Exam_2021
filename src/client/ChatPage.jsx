import React, { useEffect } from 'react';
import ReactDOM from 'react-dom';
import { useState } from 'react';


export function ChatPage() {
  const [chatLog, setChatLog] = useState([]);
  const [message, setMessage] = useState("");
  const [ws, setWs] = useState();
  
  useEffect(() => {
    const ws = new WebSocket('ws://localhost:3000');
  ws.onopen = (event) => {
    console.log('opened', event);
  }

  ws.onmessage = (event) => {
    console.log('from server', event);
    setChatLog((chatLog) => [...chatLog, event.data]);
  }

  ws.onclose = (event) => {
    console.log('closed', event); 
  }
  setWs(ws);
  
  }, [])

  function handleSubmitMessage(e) {
    e.preventDefault();
    
    ws.send(message);
    setMessage("");
  }


  return <>
  <header><h1>Chat Page</h1></header>

  <main>
    <div id="chatLog">
    {chatLog.map((message, index) => (
    <div key={index}>{message}</div>
    ))}
    </div>
  </main>
  <footer> 
    <form onSubmit={handleSubmitMessage}>
    <input type="text" autoFocus={true} value={message} onChange={e => setMessage(e.target.value)}/>
    <button>Submit</button>
    </form>
  </footer>
  </>
}