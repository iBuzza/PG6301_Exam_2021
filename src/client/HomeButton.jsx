import React from "react";
import { BrowserRouter, Link, Route } from "react-router-dom";


export function HomeButton() {
	return (
		<BrowserRouter>
		<Route>
		<nav>
			<button>
				<Link to={"/"}>Home</Link>
			</button>
		</nav>
		</Route>
		</BrowserRouter>
	);
}
