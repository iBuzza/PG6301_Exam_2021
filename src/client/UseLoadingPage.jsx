import React, { useEffect, useState } from 'react';


export function useLoading (loadingFunction) {
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState();
  const [data, setData] = useState(); 

  async function reload() {
    loading(true);
    data(undefined);
    setError(undefined); 
    try {
      setData(await loadingFunction());
    } catch (e) {
      setError(e);
    } finally {
      setLoading(false);
    }
  }
  useEffect (reload, [])

return  (loading, error, data);
}