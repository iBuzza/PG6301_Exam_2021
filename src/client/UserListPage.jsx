import React, { useEffect, useState } from 'react';

export function UserListPage() {
  const [users, setUsers] = useState();
  const [error, setError] = useState();

  async function loadUsers() {
    try {
      const res = await fetch('/api/users');
      if (!res.ok) {
        throw new Error(`Something went wrong loading ${res.url}: ${res.statusText}`);
      }

      const json = await res.json();
      setUsers(json);
    } catch (e) {
      setError(e);
    }
  }

  useEffect(loadUsers, []);

  if (error) {
    return <div>Something went wrong</div>;
  }

  if (!users) {
    return <div>...Loading...</div>;
  }

  return <>
    <h1>List users</h1>
    {users.map(({id, firstname, lastname, email}) => (
      <ul key={id}>
        <li key={id}>First name: {firstname}</li>
        <li key={id}>Last name: {lastname}</li>
        <li key={id}>Email: {email}</li>
      </ul>
    ))}
  </>;
}