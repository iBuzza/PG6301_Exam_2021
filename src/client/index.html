<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>PGR6301 Exam 2021</title>
</head>
<body>
  <div id="root"></div>
</body>
<script src="index.jsx"></script>
</html>