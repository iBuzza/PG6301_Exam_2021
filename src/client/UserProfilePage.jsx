import React from 'react';
import { useLoading } from './UseLoadingPage';
import { fetchJson } from './http';



  
function ErrorView(error ) {
  return <div>An error occured:  {error.toString()}</div>
}

function LoadingView() {
  return <div>Loading...</div>;
}

export function UserProfile() {
  const [loading, error, data ] = useLoading(() => fetchJson('/api/profile'));

  if(loading || !data) {
    return <LoadingView />;
  }

  if(Error) {
    return <ErrorView error={error}/> 
  }

  const {username} = data; 

  return <div>
    <h1>Username: {username}</h1>
  </div>;
}