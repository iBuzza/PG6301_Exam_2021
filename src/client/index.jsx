import React from "react";
import ReactDOM from "react-dom";
import { BrowserRouter, Link, Route, Switch } from "react-router-dom";
import { HomeButton } from "./HomeButton";
import { UserListPage } from "./UserListPage";
import { UserRegister } from "./UserRegisterPage";

function Application() {
	return (
		<BrowserRouter>
			<HomeButton />
			<Switch>
				<Route path={"/chat"}>
					<h1>Chatroom</h1>
				</Route>
				<Route path={"/register"}>
					<UserRegister />
				</Route>
				<Route path={"/userlist"}>
					<UserListPage />
				</Route>
				<Route exact path={"/"}>
					<h1>Chat Application</h1>
					<ul>
						<li>
							<Link to={"/chat"}>Enter Chatroom</Link>
						</li>
						<li>
							<Link to={"/register"}>Register a new user</Link>
						</li>
						<li>
							<Link to={"/userlist"}>List of users</Link>
						</li>
					</ul>
				</Route>
			</Switch>
		</BrowserRouter>
	);
}

ReactDOM.render(<Application />, document.getElementById("root"));
