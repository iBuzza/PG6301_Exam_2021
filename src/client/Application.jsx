import React from "react";
import ReactDOM from "react-dom";
import { BrowserRouter, Link, Route, Switch } from "react-router-dom";
import { ChatPage } from "./ChatPage";
import { HomeButton } from "./HomeButton";
import { UserListPage } from "./UserListPage";
import { UserRegister } from "./UserRegisterPage";
import { LoginPage } from "./LoginPage";
import { UserProfile } from './UserProfilePage';
import './styles/chatPageStyles.css';

export function Application() {
	return (
		<BrowserRouter>
			<HomeButton />
			<Switch>
      <Route path="/login">
          <LoginPage />
        </Route>
				<Route path="/profile">
          <UserProfile />
        </Route>
				<Route path={"/chat"}>
					<ChatPage />
				</Route>
				<Route path={"/register"}>
					<UserRegister />
				</Route>
				<Route path={"/userlist"}>
					<UserListPage />
				</Route>
				<Route exact path={"/"}>
					<h1>PG6301 Exam</h1>
					<ul>
            <li>
							<Link to={"/login"}>Login</Link>
						</li>
						<li>
							<Link to={"/profile"}>Profile</Link>
						</li>
						<li>
							<Link to={"/chat"}>Enter Chatroom</Link>
						</li>
						<li>
							<Link to={"/register"}>Register a new user</Link>
						</li>
						<li>
							<Link to={"/userlist"}>Users</Link>
						</li>
					</ul>
				</Route>
			</Switch>
		</BrowserRouter>
	);
}