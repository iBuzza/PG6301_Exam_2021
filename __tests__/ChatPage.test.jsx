import { ChatPage } from '../src/client/ChatPage';
import TestRenderer from 'react-test-renderer';

import React from "react";
import ReactDOM from "react-dom";
import  { act} from "react-dom/test-utils"

describe("loading view", () => {
    it("show loading view", () => {
        const view = TestRenderer.create(<ChatPage />);
        expect(view.toJSON()).toMatchSnapshot();
    });
    
    it("show loading view on dom", () => {
        const container = document.createElement("div");
        document.body.appendChild(container);
        act(() => {
            ReactDOM.render(<ChatPage />, container)
        });
        
        expect(container.innerHTML).toMatchSnapshot();
    })
    
})