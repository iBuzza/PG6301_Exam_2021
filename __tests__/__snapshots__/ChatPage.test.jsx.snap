// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loading view show loading view 1`] = `
Array [
  <header>
    <h1>
      Chat Page
    </h1>
  </header>,
  <main>
    <div
      id="chatLog"
    />
  </main>,
  <footer>
    <form
      onSubmit={[Function]}
    >
      <input
        autoFocus={true}
        onChange={[Function]}
        type="text"
        value=""
      />
      <button>
        Submit
      </button>
    </form>
  </footer>,
]
`;

exports[`loading view show loading view on dom 1`] = `"<header><h1>Chat Page</h1></header><main><div id=\\"chatLog\\"></div></main><footer><form><input type=\\"text\\" value=\\"\\"><button>Submit</button></form></footer>"`;
